akka-cluster-tools/src/main/scala/akka/cluster/pubsub/DistributedPubSubMediator.scala

```diff
@@ -258,6 +290,47 @@ object DistributedPubSubMediator {
 
     def encName(s: String) = URLEncoder.encode(s, "utf-8")
 
+    def mkKey(ref: ActorRef): String = mkKey(ref.path)
+
+    def mkKey(path: ActorPath): String = path.toStringWithoutAddress
+
+    trait PerGroupingBuffer {
+      private type BufferedMessages = Vector[(Any, ActorRef)]
+
+      private var buffers: Map[String, BufferedMessages] = Map.empty
+
+      private var totalBufferSize = 0
+
+      def bufferOr(limit: Int, grouping: String, message: Any, originalSender: ActorRef, discard: Any ⇒ Unit)(action: ⇒ Unit): Unit = {
+        buffers.get(grouping) match {
+          case None ⇒ action
+          case Some(messages) if totalBufferSize < limit ⇒
+            buffers = buffers.updated(grouping, messages :+ ((message, originalSender)))
+            totalBufferSize += 1
+          case Some(_) ⇒ discard(message)
+        }
+      }
+
+      def recreateAndForwardMessagesIfNeeded(grouping: String, recipient: ⇒ ActorRef): Unit = {
+        buffers.get(grouping).filter(_.nonEmpty).foreach(forwardMessages(_, recipient))
```

Comment:
this method is called when Terminate message finally arrives to the Mediator/Topic. When that happens if there is grouping and there are messages buffered we have to recreate Topic/Group actor and forward the messages so `.filter(_nonEmpty)` return Some(messages) only if there is at least one message buffered. 

If there is no grouping or we did not buffer any messages we do not have to recreate the actor. This solves the problem mentioned in previous "iteration" - that only checking if there is a grouping for given key is not enough in case when we want to make sure we do not have to recreate child actor.